lib/ocaml_delimiter.ml

```diff
@@ -0,0 +1,105 @@
+(*
+ * Copyright (c) 2018 Thomas Gazagnaire <thomas@gazagnaire.org>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+open Result
+
+type syntax = Cmt | Attr
+
+type part_begin = { indent : string; payload : string }
+
+type t = Part_begin of syntax * part_begin | Part_end
+
+module Regexp = struct
+  let marker = Re.str "$MDX"
+
+  let spaces = Re.rep1 Re.space
+
+  let id = Re.(rep1 (alt [ alnum; char '_'; char '-' ]))
+
+  let cmt_assign =
+    let open Re in
+    compile
+    @@ seq
+         [
+           group (rep space);
+           str "(*";
+           spaces;
+           marker;
+           spaces;
+           group id;
+           str "=";
+           group id;
+           spaces;
+           str "*)";
+         ]
+
+  let cmt_simple =
+    let open Re in
+    compile
+    @@ seq [ str "(*"; spaces; marker; spaces; group id; spaces; str "*)" ]
+
+  let attribute =
+    let open Re in
+    let ws = rep space in
+    compile @@ whole_string
+    @@ seq
+         [
+           group ws;
+           str "[@@@";
+           ws;
+           group id;
+           ws;
+           str "\"";
+           group id;
+           str "\"";
+           ws;
+           str "]";
+           ws;
+           opt (str ";;");
+           ws;
+         ]
+
+  let matches regexp line =
+    match Re.exec_opt regexp line with
+    | Some g ->
+        let indent = Re.Group.get g 1 in
+        let name = Re.Group.get g 2 in
+        let payload = Re.Group.get g 3 in
+        Some (indent, name, payload)
+    | None -> None
+end
+
+let parse line =
+  let error () =
+    Util.Result.errorf "'%s' is not a valid ocaml delimiter for mdx" line
+  in
+  match Regexp.matches Regexp.attribute line with
```

Comment:
Hmm I had something a bit more straightforward in mind.

For instance the parsing of the attribute is pretty simple, it's either a delimiter attribute or not, there can be no error here so we could extract a `parse_attr : string -> t option`.
For the comments it's a bit different but I think it makes sense to have a single function
``parse_cmt : string -> (t option, [`Msg of string]) result`` so that the part that determines whether it's an MDX comment and returns the indent, the name and the payload is shared.

I think this could result in a more robust parser.

From a highlevel point of view it would look like:
```ocaml
match parse_attr line with
| Some _ as delimiter-> Ok delimiter
| None -> parse_cmt line
```

`parse_cmt` would return `None` if we're not looking at an MDX comment and an error if it does start with `(* $MDX` but the pair `name` and `payload` don't fit together or don't make sense to MDX. Does that make sense?

What I have in mind is something like:
```ocaml
let parse_cmt line =
  match Re.exec_opt Regexp.cmt_simple with
  | Some g ->
    match Re.Group.get g 1 with
    |  "part-end" -> Ok (Some Part_end)
    | s when String.starts_with "part-begin=" -> Ok (Some (Part_begin (Cmt, { indent; payload = ...})))
    | _ -> error ()
  | None -> Ok None
```

I took some shortcut here but you get the idea. What would you think about such a solution?
I think it simplifies this bit quite a lot and in the future `parse` will just become `parse_cmt`.

Reply:
It's better indeed, done.